File: examples/web/src/news/articleIndex.ts
import type { Article } from "@aetherfeed/domain";
import { loadEncryptedJson, saveEncryptedJson } from "../lock/cache";

export const ARTICLE_INDEX_KEY = "af-lock-article-index";

export function parseArticleIndex(raw: unknown): Article[] {
  if (!Array.isArray(raw)) return [];
  return raw.flatMap((row) => {
    if (!row || typeof row !== "object") return [];
    const rec = row as Record<string, unknown>;
    if (typeof rec.id !== "string" || !rec.id.trim()) return [];
    if (typeof rec.feedId !== "string" || !rec.feedId.trim()) return [];
    if (typeof rec.title !== "string" || typeof rec.url !== "string") return [];
    return [
      {
        id: rec.id,
        feedId: rec.feedId,
        title: rec.title,
        url: rec.url,
        publishedAt: typeof rec.publishedAt === "number" ? rec.publishedAt : undefined,
        summary: typeof rec.summary === "string" ? rec.summary : undefined,
        contentHtml: typeof rec.contentHtml === "string" ? rec.contentHtml : undefined,
      },
    ];
  });
}

export async function loadArticleIndex(): Promise<Article[]> {
  return parseArticleIndex(await loadEncryptedJson<unknown>(ARTICLE_INDEX_KEY, []));
}

export async function saveArticleIndex(rows: Article[]): Promise<void> {
  await saveEncryptedJson(ARTICLE_INDEX_KEY, rows);
}

File: examples/android/app/src/main/java/org/aetherfeed/app/news/ArticleIndex.kt
package org.aetherfeed.app.news

import org.aetherfeed.app.applock.EncryptedCache
import org.aetherfeed.app.domain.Article
import org.json.JSONArray
import org.json.JSONObject

const val ARTICLE_INDEX_KIND = "article-index"

interface ArticleIndexStore {
    fun load(feedId: String): List<Article>

    fun save(feedId: String, rows: List<Article>)
}

object NoopArticleIndex : ArticleIndexStore {
    override fun load(feedId: String): List<Article> = emptyList()

    override fun save(feedId: String, rows: List<Article>) = Unit
}

fun encodeArticleIndex(rows: List<Article>): String {
    val arr = JSONArray()
    for (row in rows) {
        arr.put(
            JSONObject().apply {
                put("id", row.id)
                put("feedId", row.feedId)
                put("title", row.title)
                put("url", row.url)
                put("publishedAt", row.publishedAt ?: JSONObject.NULL)
                put("summary", row.summary ?: JSONObject.NULL)
                put("contentHtml", row.contentHtml ?: JSONObject.NULL)
            },
        )
    }
    return arr.toString()
}

fun decodeArticleIndex(raw: String): List<Article> {
    if (raw.isBlank()) return emptyList()
    val arr = runCatching { JSONArray(raw) }.getOrNull() ?: return emptyList()
    return (0 until arr.length()).mapNotNull { index ->
        val obj = arr.optJSONObject(index) ?: return@mapNotNull null
        val id = obj.optString("id")
        val feedId = obj.optString("feedId")
        if (id.isBlank() || feedId.isBlank()) return@mapNotNull null
        Article(
            id = id,
            feedId = feedId,
            title = obj.optString("title"),
            url = obj.optString("url"),
            publishedAt = if (obj.isNull("publishedAt")) null else obj.optLong("publishedAt"),
            summary = obj.optString("summary").ifBlank { null },
            contentHtml = obj.optString("contentHtml").ifBlank { null },
        )
    }
}

class EncryptedArticleIndex(private val cache: EncryptedCache) : ArticleIndexStore {
    override fun load(feedId: String): List<Article> {
        val raw = cache.read(ARTICLE_INDEX_KIND, feedId)?.toString(Charsets.UTF_8) ?: return emptyList()
        return runCatching { decodeArticleIndex(raw) }.getOrDefault(emptyList())
    }

    override fun save(feedId: String, rows: List<Article>) {
        runCatching { cache.write(ARTICLE_INDEX_KIND, feedId, encodeArticleIndex(rows).toByteArray()) }
    }
}
